plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Transform/Inline.hs

```diff
@@ -326,14 +326,19 @@ sizeIsAcceptable = \case
   Delay _ t  -> sizeIsAcceptable t
 
 isPure :: Term name uni fun a -> Bool
-isPure = go
+isPure = go True
     where
-        go = \case
-            Var {}      -> True
+        go delayIsPure = \case
+            Var {}                        -> True
             -- These are syntactically values that won't reduce further
-            LamAbs {}   -> True
-            Constant {} -> True
-            Delay {}    -> True
-
+            LamAbs {}                     -> True
+            Constant {}                   -> True
+            Delay _ body                  -> delayIsPure || go delayIsPure body
+            Apply _ (LamAbs _ _ body) arg -> go delayIsPure arg && go delayIsPure body
```

Comment:
Okay, so we're not doing it in PIR, but we are doing it here? Why? Presumably because we don't have let-bindings here so we need it in order to spot things like that. But that definitely needs a comment. And TBH that seems to me like an argument for doing it in PIR too - just consistency.

Reply:
This is actually an instance where I think it would have been helpful to do both in one PR, so we can spot whether they're consistent. Keeping the PIR and UPLC inliners in sync is a pain, best to update them together.